### DISCUSSION TITLE
Explainers, Pedagogy, and Notebooks

### DISCUSSION THREAD

**Discussion by bnord** [deepskies/DeepWiki#29] [Ideas]
(Astro)physicists are accustomed to having notebooks that can be run right out of the gate when they pick up a new project or code base. Also, the short time frames of internships and other activities make it important that users can figure out how to use code quickly; often, having a jupyter notebook that runs without issues can serve these purposes.

However, it appears that jupyter is a heavyweight dependency. 

is there a way to allow for this dependency (in our fastsim repo's, for example) despite its heaviness? 

is there another we can use that will allow people to run something immediately and get the visualizations? 

I think the HTML versions of the notebooks may not suffice.

**Comment by voetberg** [comment]
Notebooks can be included without using jupyter as a dependency. If someone is installing a package, they still need to run a kernel to start a notebook, and if notebooks are as universal as you describe, I think it's a reasonable expectation that they know how to start a kernel without it being installed automatically. 

An alternate is to include a link to a colab notebook, or instructions to open the code in colab.

**Comment by voetberg** [comment]
I do want to hear your justification why showing examples is insufficient. Is there a wider expectation that people won't know how to start up code on their own? Or is the standard for people to always install from source and do their work within the source package instead of adding to a larger package?

**Reply by bnord** [reply]
I think physicists (especially students) are much more likely to start using a package the less they need to do to prove to themselves that it works well.

**Reply by bnord** [reply]
Especially folks/students/early-career scientists who don't have a background in using a package manager or installer.

**Reply by bnord** [reply]
That's why our colleagues at space telescope made this: https://archive.stsci.edu/hello-universe

**Reply by bnord** [reply]
Or maybe if there were a way for people to load the HTML into google colab and run it, perhaps that would suffice.

**Reply by voetberg** [reply]
In that case, including a notebook (but not having jupyter as an install dependency), is totally doable. Instructions to copy to colab works (just direct people to [this link](https://colab.research.google.com/github/)), or simply stating "To open our example notebooks, run `pip install jupyter; jupyter-notebook ./notebooks`" also work

**Reply by voetberg** [reply]
I suppose the main reason I push back on saying "Just give people an example with a jupyter notebook" is two fold; 1. It is a crutch people can rely on instead of writing good documentation. It limits what people can learn to do with your code down the road, 2. You can't properly develop in jupyter notebooks. If we want to install good practices, in both our group and the community as a whole,  we should try to encourage people to write code that is designed to be used by other people, outside the narrow scope of their own projects.

**Reply by bnord** [reply]
I agree with you on both your points. I'm trying to either a) split the difference or b) have it both ways -- I'm not sure which, though, lol.

I'm leaning into to making something as easy as possible for people to adopt AND saying that it's up to the user to use it properly.

**Reply by voetberg** [reply]
I think directing people to colab is probably the closest to split the difference we can then. Perhaps there should be a policy of linking to documentation when talking about more advanced behavior, and keeping the notebook for exclusively simple solutions (e.g. Import the package, execute the simplest version of the code).

**Comment by voetberg** [comment]
I think I have a solution to this, after having to work a little in colab for running tutorials. 

When making a documentation page, it is simple enough to add a 'open in colab' button at the top of an embedded HTML page. 

This would look having a documentation rst page that looks like: 

```
Example
---------

.. image:: https://colab.research.google.com/assets/colab-badge.svg 
   :target: https://colab.research.google.com/github/{path to the example notebook}

.. raw:: html
   :file: examples/{example notebook exported}.html

```

Which produces a page that looks like this: 

![image](https://github.com/deepskies/DeepWiki/assets/27649232/2b94c2d8-0e25-4616-a8de-5ddff1c7a2ab)
